Spigot-Server-Patches/0553-NamespacedKey-based-plugin-chunk-tickets.patch

```diff
@@ -0,0 +1,267 @@
++                ticket -> namespace.equals(ticket.getObjectReason().getNamespace()));
++    }
++
++    @Override
++    public Collection<io.papermc.paper.TickingLevel> getKeyedPluginChunkTicketLevels(int x, int z, org.bukkit.NamespacedKey key) {
++        Preconditions.checkArgument(key != null, "null key");
++
++        // logic copied from getPluginChunkTickets
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        Collection<Ticket<?>> tickets = chunkDistanceManager.tickets.get(ChunkCoordIntPair.pair(x, z));
++        if (tickets == null) {
++            return Collections.emptyList();
++        }
++
++        ImmutableList.Builder<io.papermc.paper.TickingLevel> ret = ImmutableList.builder();
++        for (Ticket<?> ticket : tickets) {
++            if (ticket.getTicketType() == TicketType.KEYED_PLUGIN_TICKET && key.equals(ticket.getObjectReason())) {
++                ret.add(io.papermc.paper.PaperTickingLevel.getTickingLevel(ticket.getTicketLevel()));
++            }
++        }
++        return ret.build();
++    }
++
++    @Override
++    public Collection<org.bukkit.NamespacedKey> getKeyedPluginChunkTickets(int x, int z) {
++        // logic copied from getPluginChunkTickets
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++        Collection<Ticket<?>> tickets = chunkDistanceManager.tickets.get(ChunkCoordIntPair.pair(x, z));
++        if (tickets == null) {
++            return Collections.emptyList();
++        }
++
++        ImmutableList.Builder<org.bukkit.NamespacedKey> ret = ImmutableList.builder();
++        for (Ticket<?> ticket : tickets) {
++            if (ticket.getTicketType() == TicketType.KEYED_PLUGIN_TICKET) {
++                ret.add((org.bukkit.NamespacedKey) ticket.getObjectReason());
++            }
++        }
++        return ret.build();
++    }
++
++    @Override
++    public Map<org.bukkit.NamespacedKey, Collection<Chunk>> getKeyedPluginChunkTickets() {
++        // logic copied from getPluginChunkTickets
++        Map<org.bukkit.NamespacedKey, ImmutableList.Builder<Chunk>> ret = new HashMap<>();
++        ChunkMapDistance chunkDistanceManager = this.world.getChunkProvider().playerChunkMap.chunkDistanceManager;
++
++        for (Long2ObjectMap.Entry<ArraySetSorted<Ticket<?>>> chunkTickets : chunkDistanceManager.tickets.long2ObjectEntrySet()) {
++            long chunkKey = chunkTickets.getLongKey();
++            Collection<Ticket<?>> tickets = chunkTickets.getValue();
++
++            Chunk chunk = null;
++            for (Ticket<?> ticket : tickets) {
++                if (ticket.getTicketType() != TicketType.KEYED_PLUGIN_TICKET) {
++                    continue;
++                }
++
++                if (chunk == null) {
++                    chunk = this.getChunkAt(ChunkCoordIntPair.getX(chunkKey), ChunkCoordIntPair.getZ(chunkKey));
```

Comment:
this will promote every chunk to entity ticking due to this call.
This will need to instead look up the Chunk object in nms and use chunk.bukkitChunk instead